### DISCUSSION TITLE
Jobs with time based attempts that fail when pushed to queue

### DISCUSSION THREAD

**Discussion by mikkoaf** [laravel/framework#44583] [Ideas]
# What

It's possible to dispatch jobs that will fail without a single successful (or failing due to other factors) try due to a time based attempts check. This can happen either by setting up the `retryUntil` property into the past, or by setting up a `retryUntil` to a time sooner than what the job is delayed.

https://laravel.com/docs/9.x/queues#delayed-dispatching
https://laravel.com/docs/9.x/queues#time-based-attempts

# Why

A new exception could be thrown checking to ensure that a delayed or pushed job with time based attempts can be run at least once. It would make the feedback loop of catching this particular type of issues quicker during development and in production. 

# My suggestion

I've come up with a prototype, https://github.com/mikkoaf/framework/tree/available-after-retry-until 
I've already covered that closures and jobs that return no expiration time will pass, and it is implicitly tested with the existing testcases using strings as jobs.

However, I am hesitant to create a pull request just yet as I have to figure out a few things.

- The current implementation works just fine, why bother? Having this feature initially would have saved me just only a minute when I first encountered this issue.
- This change could have major impact in complex job chains and rate limited job dispatching.
- The prototype throws an exception instead of allowing the job to fail as it has been configured, with the job's own `fail()` method running.
   - Instead of throwing an exception, should the job be allowed to fail with it's own fail method?
   - Would this change how for rate limited processes are handled if a specific job failure would run?
- The tests and implementation has repetition.
  - I initially looked in to implementing this at the start of https://github.com/laravel/framework/blob/8f82cf970479f4949c6cf76f89cec01ee9f2e252/src/Illuminate/Queue/Queue.php#L302 but as Queue fake facade does not use the abstraction, I thought I would seek another solution.

**Comment by alejandrovelez7** [comment]
Running into the same issue in 2024. I'd love a different exception for this or for it to somehow be reflected in the JobFailed event at least.
